runtime/bindings/python/iree/runtime/io.py

```diff
@@ -64,3 +67,162 @@ def save_archive_file(entries: dict[str, Union[Any, SplatValue]], file_path: Pat
+) -> np.ndarray:
+    """Returns a tensor viewed with appropriate shape/dtype from metadata.
+
+    Raises a ValueError if unsupported.
+    """
+
+    # Decode metadata.
+    versioned_metadata = index_entry.metadata.decode()
+    metadata_parts = versioned_metadata.split(_metadata_version_separator, maxsplit=1)
+    if len(metadata_parts) == 1:
+        raise ValueError(
+            (
+                f'Invalid metadata for parameter index entry "{index_entry.key}".'
+                f' Expected format version prefix not found in "{metadata_parts[0][:100]}".'
+            )
+        )
+    format_version = metadata_parts[0]
+    metadata = metadata_parts[1]
+    if (
+        format_version != _metadata_version
+        and format_version != _metadata_iree_turbine_version
+    ):
+        raise ValueError(
+            (
+                f'Unsupported metadata format version "{format_version}" for parameter '
+                'index entry "{index_entry.key}": Cannot convert to tensor'
+            )
+        )
+    d = json.loads(metadata)
+    try:
+        type_name = d["type"]
+        if d["type"] != "Tensor":
+            raise ValueError(
+                f"Metadata for parameter entry {index_entry.key} is not a Tensor ('{type_name}')"
+            )
+        dtype_name = d["dtype"]
+        shape = d["shape"]
+    except KeyError as e:
+        raise ValueError(f"Bad metadata for parameter entry {index_entry.key}") from e
+
+    # Unpack/validate.
+    try:
+        dtype = _NAME_TO_DTYPE[dtype_name]
+    except KeyError:
+        raise ValueError(f"Unknown dtype name '{dtype_name}'")
+    try:
+        shape = [int(d) for d in shape]
+    except ValueError as e:
+        raise ValueError(f"Illegal shape for parameter entry {index_entry.key}") from e
+
+    t = parameter_index_entry_as_numpy_flat_ndarray(index_entry)
+    return t.view(dtype=dtype).reshape(shape)
+
+
+_DTYPE_TO_NAME = (
+    (np.float16, "float16"),
+    (np.float32, "float32"),
+    (np.float64, "float64"),
+    (np.int32, "int32"),
+    (np.int64, "int64"),
+    (np.int16, "int16"),
+    (np.int8, "int8"),
+    (np.uint32, "uint32"),
+    (np.uint64, "uint64"),
+    (np.uint16, "uint16"),
+    (np.uint8, "uint8"),
+    (np.bool_, "bool"),
+    (np.complex64, "complex64"),
+    (np.complex128, "complex128"),
+)
+
+_NAME_TO_DTYPE: dict[str, np.dtype] = {
+    name: np_dtype for np_dtype, name in _DTYPE_TO_NAME
+}
+
+
+def _map_dtype_to_name(dtype) -> str:
+    for match_dtype, dtype_name in _DTYPE_TO_NAME:
+        if match_dtype == dtype:
+            return dtype_name
+
+    raise KeyError(f"Numpy dtype {dtype} not found.")
+
+
+_metadata_version = "TENSORv0"
+"""Magic number to identify the format version.
+The current version that will be used when adding tensors to a parameter index."""
+
+_metadata_iree_turbine_version = "PYTORCH"
+"""There are files created with IREE Turbine that use this prefix.
+This is here to maintain the ability to load such files."""
```

Comment:
(Feel free to merge without this, but it would be nice)

We can probably trust this code... but if you want to go a bit further and guard against compatibility breaking changes, you could add a unit test with some binary data exported from iree-turbine to ensure that it is supported here:

```python
def testParameterIndexEntryFromIreeTurbine(self):
    index = rt.ParameterIndex()
    index.load("test_iree_turbine_index.irpa")  # load from a testdata file
    index_entry_as_array = rt.parameter_index_entry_as_numpy_ndarray(...)
    # then do some sanity check on the returned ndarray
```

(avoiding the temptation to put logic in the test and call the `_make_tensor_metadata` helper function: https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html)

Reply:
Nice, thanks!